packages/pds/src/sequencer/outbox.ts

```diff
@@ -0,0 +1,111 @@
+import { AsyncBuffer, AsyncBufferFullError } from '@atproto/common'
+import Sequencer, { RepoAppendEvent } from '.'
+
+export type OutboxOpts = {
+  maxBufferSize: number
+}
+
+export class Outbox {
+  caughtUp = false
+  lastSeen = -1
+
+  cutoverBuffer: RepoAppendEvent[]
+  outBuffer: AsyncBuffer<RepoAppendEvent>
+
+  constructor(public sequencer: Sequencer, opts: Partial<OutboxOpts> = {}) {
+    const { maxBufferSize = 500 } = opts
+    this.cutoverBuffer = []
+    this.outBuffer = new AsyncBuffer<RepoAppendEvent>(maxBufferSize)
+  }
+
+  // event stream occurs in 3 phases
+  // 1. backfill events: events that have been added to the DB since the last time a connection was open.
+  // The outbox is not yet listening for new events from the sequencer
+  // 2. cutover: the outbox has caught up with where the sequencer purports to be,
+  // but the sequencer might already be halfway through sending out a round of updates.
+  // Therefore, we start accepting the sequencer's events in a buffer, while making our own request to the
+  // database to ensure we're caught up. We then dedupe the query & the buffer & stream the events in order
+  // 3. streaming: we're all caught up on historic state, so the sequencer outputs events and we
+  // immediately yield them
+  async *events(backfillFrom?: string): AsyncGenerator<RepoAppendEvent> {
+    // catch up as much as we can
+    if (backfillFrom) {
+      for await (const evt of this.getBackfill(backfillFrom)) {
+        yield evt
+        this.lastSeen = evt.seq
```

Comment:
Similar question here: should consumers of the outbox see the event before or after the outbox itself?  Here's an example:
```js
for await (const ev of outbox.events()) {
  assert(ev.seq <= outbox.lastSeen)
}
```